<a href="https://colab.research.google.com/github/rsanchezobreque-stack/Supermercado-RS-nchez/blob/main/Lista_de_ventas_de_un_supermercado%2C_tarea_14_11.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Raúl Sánchez
Maximiliano Gonzalez
Ignacion Ulloa

1. Producto Mas Caro

In [ ]:
from google.colab import files      # Importa la herramienta para subir archivos en Google Colab
import pandas as pd                 # Importa pandas para leer y trabajar con datos

print("Sube el archivo productos.csv")  # Mensaje para pedir que suban el archivo
uploaded = files.upload()                # Abre el cuadro para subir el archivo

# Nombre del archivo subido
archivo = list(uploaded.keys())[0]       # Obtiene el nombre del archivo que subiste

# Leer el archivo sin encabezados (porque tu archivo NO tiene títulos)
df = pd.read_csv(archivo, header=None)   # Lee el CSV como tabla

# Columna 1: nombre del producto
# Columna 2: precio del producto
col_nombre = 1                            # Indica que la columna 1 es el nombre
col_precio = 2                            # Indica que la columna 2 es el precio

# Convertir la columna de precios a número (por si viene como texto)
df[col_precio] = pd.to_numeric(df[col_precio], errors="coerce")  # Convierte a número y pone NaN si hay error

# Buscar la fila donde el precio es máximo (producto más caro)
fila_mas_cara = df.loc[df[col_precio].idxmax()]   # idxmax() da la posición del precio más alto

# Mostrar resultados
print("\nEl producto más caro es:", fila_mas_cara[col_nombre])  # Imprime el nombre del producto más caro
print("Con un precio de:", fila_mas_cara[col_precio])           # Imprime el precio del producto más caro


Sube el archivo productos.csv


TypeError: 'NoneType' object is not subscriptable

2. Valor total de la **bodega**

In [ ]:
# Importamos el módulo csv para poder leer archivos .csv fácilmente
import csv

# Definimos una funcion llamada "valorTotalBodega" que calcula el valor total de la bodega
def valorTotalBodega(archivo):
  # Inicializamos el total en 0
    total = 0
      # Creamos un lector que recorre el archivo línea por línea
    lector = csv.reader(archivo)
    for fila in lector:
        # Extraemos el precio y la cantidad desde la fila (posición 2 y 3)
        precio = int(fila[2])
        cantidad = int(fila[3])
        # Sumamos al total el valor de ese producto (precio * cantidad)
        total += precio * cantidad
         # Devolvemos el total acumulado
    return total

# Abrimos el archivo productos.csv en modo lectura
with open('productos.csv', 'r') as productos:
    # Llamamos a la función y mostramos el resultado en pantalla
    print(f"La bodega vale en total {valorTotalBodega(productos)}")


La bodega vale en total 1900570


3. Producto Con Mas Ingresos

In [ ]:
ITEMS_PATH = "items.csv"
PRODUCTOS_PATH = "productos.csv"

def leer_lineas(path):
    with open(path, "r", encoding="utf-8") as f:
        lineas = f.readlines()
    # limpiar retornos de carro innecesarios
    lineas = [l.rstrip("\n\r") for l in lineas]
    # eliminar líneas vacías
    lineas = [l for l in lineas if l.strip() != ""]
    return lineas

def separar_fila(linea):
    #Separa una línea por ';' o ',' según corresponda (prioriza ';' si existe)
    if ";" in linea:
        return [c.strip() for c in linea.split(";")]
    else:
        return [c.strip() for c in linea.split(",")]

#eer archivos como en tus ejemplos ---
lineas_items = leer_lineas(ITEMS_PATH)
lineas_productos = leer_lineas(PRODUCTOS_PATH)

# Convertir a listas de listas (como en tus ejemplos con .split)
rows_items = [separar_fila(l) for l in lineas_items]
rows_productos = [separar_fila(l) for l in lineas_productos]

#Detectar si hay cabeceras (usando la técnica vista: si primer campo no es dígito)
start_items = 0
if rows_items and not rows_items[0][0].isdigit():
    start_items = 1

start_prod = 0
if rows_productos and not rows_productos[0][0].isdigit():
    start_prod = 1

#Construir diccionario id -> (nombre, precio) usando solo lo enseñado
info_productos = {}
for fila in rows_productos[start_prod:]:
    #productos: id_producto ; nombre ; precio ; stock  (según lo entregado)
    if len(fila) < 3:
        continue
    id_prod = fila[0]
    nombre = fila[1]
    #convertir precio a float (si hay errores, usar 0)
    try:
        precio = float(fila[2])
    except:
        #si el precio tiene separadores de miles (p.ej. "1.200"), intentar limpieza simple:
        t = fila[2].replace(".", "").replace(",", ".")
        try:
            precio = float(t)
        except:
            precio = 0.0
    info_productos[id_prod] = (nombre, precio)

#Acumular ingresos por nombre (precio * cantidad)
ingresos = {}
for fila in rows_items[start_items:]:
    #items: id_item ; id_producto ; cantidad
    if len(fila) < 3:
        continue
    id_prod = fila[1]
    #convertir cantidad a float
    try:
        cantidad = float(fila[2])
    except:
        t = fila[2].replace(".", "").replace(",", ".")
        try:
            cantidad = float(t)
        except:
            cantidad = 0.0

    if id_prod in info_productos:
        nombre, precio = info_productos[id_prod]
        ingreso = precio * cantidad
        if nombre in ingresos:
            ingresos[nombre] += ingreso
        else:
            ingresos[nombre] = ingreso
    else:
        #Si el producto no está en el catálogo, lo ignoramos (tal como en tu enunciado).
        pass

#Resultado: producto con más ingresos
if not ingresos:
    print("No se pudieron calcular ingresos. Revisa los archivos.")
else:
    #imprimir detalle ordenado descendente
    orden = sorted(ingresos.items(), key=lambda x: x[1], reverse=True)
    print("Ingresos por producto (descendente):")
    for nombre, ingreso in orden:
        #formato: si entero sin decimales, mostrar entero; si no, 2 decimales
        if abs(ingreso - int(ingreso)) < 1e-9:
            ingreso_str = f"{int(ingreso)}"
        else:
            ingreso_str = f"{ingreso:.2f}"
        print(f" - {nombre}: {ingreso_str}")

    producto_mayor = orden[0][0]
    mayor_valor = orden[0][1]
    if abs(mayor_valor - int(mayor_valor)) < 1e-9:
        mayor_str = f"{int(mayor_valor)}"
    else:
        mayor_str = f"{mayor_valor:.2f}"

    print("\nProducto con más ingresos:")
    print(f"{producto_mayor}  →  {mayor_str}")

Ingresos por producto (descendente):
 - Arroz: 5400
 - Desodorante: 2190
 - Palta: 2000
 - Yogur: 1584
 - Bebida: 1400
 - Fideos: 900
 - Galletas: 470

Producto con más ingresos:
Arroz  →  5400


4. Total  de ventas al mes

In [ ]:
import csv

def separar(linea):
    return linea.strip().split(";")

def totalVentasMes(archivo_items, archivo_productos, archivo_ventas, mes, año):

    productos = {}
    try:
      #esto se pone por si el codigo puede fallar
        with open(archivo_productos, "r", encoding="utf-8") as f:
          #esto abre el archivo y lo cierra
            reader = csv.reader(f, delimiter=',')
            next(reader)  # saltar cabecera
            for fila in reader:
              #repite el codigo para cada cosa
                if len(fila) >= 3:

                    idProducto, _, precio = fila[:3]
                    productos[idProducto.strip()] = float(precio.strip())
                    #hace la asignacion multiple
    except FileNotFoundError:
#y con este es que se podria hacer si lo hace
        return 0
    except Exception as e:
        print(f"Error al cargar productos: {e}")
        return 0

    items_por_venta = {}
    try:
        with open(archivo_items, "r", encoding="utf-8") as f:
            next(f)
            for linea in f:
              #crea un bucle
                partes = separar(linea)
                if len(partes) >= 3:
                    idVenta, idProducto, cantidad = partes[:3]
                    if idVenta not in items_por_venta:
                        items_por_venta[idVenta] = []
                        #los if verifica que la linea tenga almenos 3 datos

                    items_por_venta[idVenta].append((idProducto, int(cantidad)))
    except FileNotFoundError:
        return 0
        #detiene la funcion, y devuelve el valor que estaba
    except Exception as e:
        print(f"Error al cargar ítems: {e}")
        return 0
    total = 0
    try:
        with open(archivo_ventas, "r", encoding="utf-8") as f:
            next(f)
            for linea in f:
                partes = separar(linea)
                if len(partes) >= 2:
                    idVenta, fecha = partes[:2]


                    try:

                        partes_fecha = fecha.split("-")

                        if len(partes_fecha) == 3:
                            dia, mesVenta, añoVenta = partes_fecha
                        else:
                            continue

                    except ValueError:
                        continue


                    if int(mesVenta) == mes and int(añoVenta) == año:
                      #sirve para verificar si la venta pertenece exactamente al mes y al año que querés consultar.

                        if idVenta in items_por_venta:
                            for idProducto, cantidad in items_por_venta[idVenta]:
                                if idProducto in productos:
                                    precio = productos[idProducto]
                                    total += precio * cantidad
    except FileNotFoundError:
        return 0
    except Exception as e:
        print(f"Error al calcular ventas: {e}")
        return 0

    return total

MES_CONSULTA = 9
AÑO_CONSULTA = 2010

resultado = totalVentasMes(
    archivo_items="items.csv",
    archivo_productos="productos.csv",
    archivo_ventas="ventas.csv",
    mes=MES_CONSULTA,
    año=AÑO_CONSULTA
)
#sin funciones que definen el total de ventas del mes y el año

print(f"El total de ventas para el mes {MES_CONSULTA:02d} del año {AÑO_CONSULTA} es: ${resultado:,.2f}")

El total de ventas para el mes 09 del año 2010 es: $0.00


 Informe
 El producto más caro es Vuvuzela
 El valor total de la bodega es de $1900570
 El producto con más ingresos es Arroz > 5400
 En el período de 09/2010, el total de ventas es de $64500